nexus/db-queries/src/db/queries/volume.rs

```diff
@@ -0,0 +1,563 @@
+    for BuildJsonResourcesToCleanUp
+{
+    type IsAggregate = is_aggregate::Never;
+}
+
+/// Produces a query fragment that will set the `resources_to_clean_up` column
+/// of the volume being deleted if it is not set already.
+///
+/// The output should look like:
+///
+/// ```sql
+///   set
+///    time_deleted = now(),
+///    resources_to_clean_up = ( select <BuildJsonResourcesToCleanUp> )
+///    where id = '<volume_id>' and
+///    (<ResourcesToCleanUpColumnIsNull>)
+///    returning volume.*
+/// ```
+#[must_use = "Queries must be executed"]
+#[derive(Debug, Clone)]
+pub struct ConditionallyUpdateVolume {
+    resources_to_clean_up_column_is_null_clause: ResourcesToCleanUpColumnIsNull,
+    build_json_resources_to_clean_up_query: BuildJsonResourcesToCleanUp,
+    volume_id: Uuid,
+}
+
+impl ConditionallyUpdateVolume {
+    pub fn new(volume_id: Uuid, table: String) -> Self {
+        Self {
+            resources_to_clean_up_column_is_null_clause:
+                ResourcesToCleanUpColumnIsNull::new(volume_id),
+            build_json_resources_to_clean_up_query:
+                BuildJsonResourcesToCleanUp::new(table, volume_id),
+            volume_id,
+        }
+    }
+}
+
+impl QueryId for ConditionallyUpdateVolume {
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl QueryFragment<Pg> for ConditionallyUpdateVolume {
+    fn walk_ast<'a>(&'a self, mut out: AstPass<'_, 'a, Pg>) -> QueryResult<()> {
+        use db::schema::volume::dsl;
+
+        let statement: String = [
+            "SET",
+            dsl::time_deleted::NAME,
+            " = now(),",
+            dsl::resources_to_clean_up::NAME,
+            "= ( SELECT ",
+        ]
+        .join(" ");
+        out.push_sql(&statement);
+
+        self.build_json_resources_to_clean_up_query.walk_ast(out.reborrow())?;
+
+        let statement: String = [
+            ")",
+            "WHERE",
+            dsl::id::NAME,
+            &format!("= '{}'", self.volume_id),
+            "AND (",
+        ]
+        .join(" ");
+        out.push_sql(&statement);
+
+        self.resources_to_clean_up_column_is_null_clause
+            .walk_ast(out.reborrow())?;
+
+        out.push_sql(") RETURNING volume.*");
+
+        Ok(())
+    }
+}
+
+impl Expression for ConditionallyUpdateVolume {
+    type SqlType = diesel::sql_types::Array<db::model::Volume>;
+}
+
+impl<GroupByClause> ValidGrouping<GroupByClause> for ConditionallyUpdateVolume {
+    type IsAggregate = is_aggregate::Never;
+}
+
+/// Produces a query fragment that will
+///
+/// 1. decrease the number of references for each region snapshot that
+///    a volume references
+/// 2. soft-delete the volume
+/// 3. record the resources to clean up as a serialized CrucibleResources
+///    struct in volume's `resources_to_clean_up` column.
+///
+/// The output should look like:
+///
+/// ```sql
+///  with UPDATED_REGION_SNAPSHOTS_TABLE as (
+///    UPDATE region_snapshot <ConditionallyDecreaseReferences>
```

Comment:
Nitpick: I think this might be more clear if the `UPDATE region_snapshot` line also lived within the `ConditionallyDecreaseReferenes` expression -- just so that the refactored expressions are like, "whole SQL queries with a knowable SQL type".

Reply:
done in d242900c197a